util/chplenv/compiler_utils.py

```diff
@@ -49,3 +64,36 @@ def CompVersion(version_string):
 def compiler_is_prgenv(compiler_val):
   return (compiler_val.startswith('cray-prgenv') or
      os.environ.get('CHPL_ORIG_TARGET_COMPILER','').startswith('cray-prgenv'))
+
+
+#
+# Determine whether a given compiler's default compilation mode
+# supports standard atomics by running the compiler and checking
+# how it expands key feature-test macros.
+#
+# The assumption is that if standard atomics make it into the
+# compiler's default compilation mode, then they actually work.
+# If they are not available in the default mode, they probably
+# have problems and we don't want to use them.
+#
+# Due to the command-line options required, this works for GCC,
+# Clang, and the Intel compiler, but probably not others.
+#
+def has_std_atomics(compiler_val):
+    try:
+        compiler_name = get_compiler_name(compiler_val)
+        if compiler_name == 'other':
+            return False
+        cmd = "echo __STDC_VERSION__ __STDC_NO_ATOMICS__ | "
+        cmd += compiler_name + " -E -x c - | sed -e '/^\#/d'"
+        out = run_command(['sh', '-c', cmd]).split()
```

Comment:
Instead of calling out to a shell, I think this would be better by adding stdin pipe support to run_command. I think this will work: 


```python
def run_command(command, stdout=True, stderr=False, cmd_input=None):

    process = subprocess.Popen(command,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE,
                               stdin=subprocess.PIPE)
    byte_cmd_input = str.encode(cmd_input) if cmd_input else None
    output = process.communicate(input=byte_cmd_input)
    if process.returncode != 0:
        raise CommandError(
            "command `{0}` failed - output was \n{1}".format(command,
                                                             output[1]))
    else:
        output = (output[0].decode(), output[1].decode())
        if stdout and stderr:
            return output
        elif stdout:
            return output[0]
        elif stderr:
            return output[1]
        else:
            return ''
```

and then your has_std_atomics function could be something like: 


```python
# strips empty and preprocessor lines
def strip_preprocessor_lines(lines):
    lines = [line for line in lines if len(line.split('#')[0].strip()) > 0]
    return lines


@memoize
def has_std_atomics(compiler_val):
    try:
        compiler_name = get_compiler_name(compiler_val)
        if compiler_name is 'other':
            return False

        version_key='version'
        atomics_key='atomics'

        cmd_input = '{0}=__STDC_VERSION__\n{1}=__STDC_NO_ATOMICS__'.format(version_key, atomics_key)
        cmd = [compiler_name, '-E', '-x', 'c', '-']
        output = run_command(cmd, cmd_input=cmd_input)
        output = strip_preprocessor_lines(output.splitlines())

        output_dict = dict(line.split('=') for line in output)
        version = output_dict[version_key].rstrip("L")
        atomics = output_dict[atomics_key]

        if version == "__STDC_VERSION__" or int(version) < 201112:
            return False
        # If the atomics macro was expanded (defined), then we do not have support.
        if atomics != "__STDC_NO_ATOMICS__":
            return False
        return True
    except:
        return False
```

It's basically what you had, except it uses pure subprocess instead of sh, strip_preprocessor_lines instead of sed, and stuffs the version and atomics into a dictionary instead of relying on the output order.